libs/cardano-ledger-core/src/Cardano/Ledger/DRepDistr.hs

```diff
@@ -0,0 +1,222 @@
+  show (DRPulsing p) = "(DRPulsing " ++ show p ++ ")"
+
+instance Crypto c => EncCBOR (DRepDistr c) where
+  encCBOR (DRComplete x) = encode (Rec DRComplete !> To x)
+  encCBOR (DRPulsing x) = encode (Rec DRComplete !> To (complete x))
+
+instance Crypto c => DecCBOR (DRepDistr c) where
+  decCBOR = decode (RecD DRComplete <! From)
+
+instance ToExpr (DRepDistr era) where
+  toExpr (DRComplete x) = App "DRComplete" [toExpr x]
+  toExpr (DRPulsing x) = App "DRComplete" [toExpr (complete x)]
+
+-- ==============================================
+-- Functions for pulsing DRepDistr
+
+startDRepDistr :: Int -> UMap c -> Set (Credential 'Voting c) -> VMap VB VP (Credential 'Staking c) (CompactForm Coin) -> DRepDistr c
+startDRepDistr _ _ regDreps _ | Set.null regDreps = DRComplete Map.empty
+startDRepDistr stepSize um regDreps stakeDistr = DRPulsing $ DRepPulser stepSize um regDreps stakeDistr Map.empty
+
+pulseDRepDistr :: DRepDistr c -> DRepDistr c
+pulseDRepDistr (DRPulsing x) = if done x2 then DRComplete (current x2) else DRPulsing x2
+  where
+    x2 = pulse x
+pulseDRepDistr x@(DRComplete _) = x
+
+completeDRepDistr :: DRepDistr c -> DRepDistr c
+completeDRepDistr (DRPulsing x) = DRComplete (complete x)
+completeDRepDistr x@(DRComplete _) = x
+
+extractDRepDistr :: DRepDistr c -> Map (DRep c) (CompactForm Coin)
+extractDRepDistr (DRPulsing x) = complete x
+extractDRepDistr (DRComplete x) = x
+
+-- ====================================================
+
+{-
+data UView c k v where
+  RewDepUView :: !(UMap c)
+                 -> UView c (Cardano.Ledger.Credential.Credential
+                               'Cardano.Ledger.Keys.Staking c) RDPair
+  PtrUView :: !(UMap c)
+              -> UView c Cardano.Ledger.Credential.Ptr (Cardano.Ledger.Credential.Credential
+                                                          'Cardano.Ledger.Keys.Staking c)
+  SPoolUView :: !(UMap c)
+                -> UView c (Cardano.Ledger.Credential.Credential
+                              'Cardano.Ledger.Keys.Staking c) (Cardano.Ledger.Keys.KeyHash
+                                                                 'Cardano.Ledger.Keys.StakePool c)
+  DRepUView :: !(UMap c)
+               -> UView c (Cardano.Ledger.Credential.Credential
+                             'Cardano.Ledger.Keys.Staking c) (DRep c)
+  	-- Defined in ‘Cardano.Ledger.UMap’
+instance Foldable (UView c k) -- Defined in ‘Cardano.Ledger.UMap’
+
+data DRep c
+  = cardano-ledger-core-1.5.0.0:Cardano.Ledger.Core.TxCert.DRepKeyHash !(KeyHash 'Voting c)
+  | cardano-ledger-core-1.5.0.0:Cardano.Ledger.Core.TxCert.DRepScriptHash !(ScriptHash c)
+  | DRepAlwaysAbstain
+  | DRepAlwaysNoConfidence
+
+data VState era
+  = VState {vsDReps :: !(Set (Credential 'Voting (EraCrypto era))),
+            vsCommitteeHotKeys :: !(Map
+                                      (KeyHash 'CommitteeColdKey (EraCrypto era))
+                                      (Maybe (Credential 'CommitteeHotKey (EraCrypto era))))
+-}
```

Comment:
Some dead code
```suggestion
```

Reply:
Forgot that was there. It was nice to refer to these things when I first got started,